### DISCUSSION TITLE
What do you need to see when mentoring a solution?

### DISCUSSION THREAD

**Discussion by iHiD** [exercism/v3#2164] [General]
**This is a discussion of the mentoring solution page, and how we want it to look/work in v3.**

**You've chosen to mentor a solution and you're giving some feedback. What needs to be on the page? What is wrong/missing in v2? What should we add in v3?**

**Note: This does _not_ cover the UX of the mentoring dashboard, just the "conversation" page.**

**All feedback _very_ gratefully received 🙂**  

---

## The student widget

Containing:
- Their handle
- Their avatar
- Their reputation
- Their responsiveness.
- If they're a favourite
- Links to previous interactions.

---

## Ability to change between iterations

A mentor needs to be able to step between iterations. It is important that mentors can clearly see which iterations they have commented on previously.

<img width="1174" alt="Screenshot 2020-08-10 at 17 22 02" src="https://user-images.githubusercontent.com/286476/89805804-170e1980-db2e-11ea-89ed-a86ce8573077.png">
<img width="1153" alt="Screenshot 2020-08-10 at 17 21 55" src="https://user-images.githubusercontent.com/286476/89805796-14abbf80-db2e-11ea-92d9-7f521d952105.png">

---

## Iteration-specific content

For each iteration, the mentor should see:
- The student's code
- The conversation on that iteration, and the ability to post a new comment.
- Any automated-suggestions
- Test results.

Some mentor's want the iteration to be full-width with comments below. Some want a left/right split. 

---

## Automated suggestions

For many solutions there will be automated suggestions which a mentor can use when mentoring the student. Also, for many (most to start with) there will not be any.

---

## Button to view mentor notes

A button to view the community-sourced mentoring notes.

<img width="888" alt="Screenshot 2020-08-10 at 17 24 31" src="https://user-images.githubusercontent.com/286476/89806003-60f6ff80-db2e-11ea-842e-e0e2fe76709d.png">
<img width="324" alt="Screenshot 2020-08-10 at 17 24 24" src="https://user-images.githubusercontent.com/286476/89805995-5e94a580-db2e-11ea-87fd-81f59fd7950a.png">
<img width="831" alt="Screenshot 2020-08-10 at 17 24 27" src="https://user-images.githubusercontent.com/286476/89805996-5f2d3c00-db2e-11ea-9ecf-dc7d5429d53c.png">

The modal should have a CTA to improve the mentoring notes. Atm, this is a footer, but many people have said it's not clear enough:
<img width="833" alt="Screenshot 2020-08-10 at 17 25 07" src="https://user-images.githubusercontent.com/286476/89806079-7d933780-db2e-11ea-9cbb-97aebc24b7bf.png">

---

## Button to private notes

Mentors keep private notes on solutions, often with snippets they can copy + paste straight to students. There should be a button that allows them to edit and interact with their notes.

---

## Button to own solution

A link to your own solution to this exercise (can open in new tab). Mentors want to remember how they solved it - seeing your own solution is the easiest way to achieve this.

---

## Button to previously mentored solutions

A link to your previous completed mentorings to this solution (this could be your mentoring dashboard filtered by this exercise and can open in new tab).

---

## Button for "no action required"

If there is no action required (for example the last comment was simply "OK") then they need a button to remove this from their workspace.

---

## Ability to download the code

Mentors need the ability to download the students code to fiddle with it locally. This currently looks like this, but should probably be a button that shows this:

<img width="557" alt="Screenshot 2020-08-10 at 17 29 04" src="https://user-images.githubusercontent.com/286476/89806495-0ad68c00-db2f-11ea-8618-4c75315d4fb5.png">

---

## Button to mark conversation as finished

Once the mentor has had enough, they should be able to mark the conversation as finished.

---

## Lock the conversation

Sometimes a mentor may get a student that turns abusive or nasty. In that situation, they should be able to lock the conversation so no more responses can be added (this probably only appears after the conversation is "finished")

**Comment by zuzia-kru** [comment]
I think that I'd use a button to see previously mentored solutions to given exercise - for now, when I see that a lot of students submit similar first iterations of the exercise I save my feedback to that in my private notes and go back to that when I see the same attempt in the future but in the beginning I tried to go back to previously mentored solutions and found it a bit awkward.

**Reply by iHiD** [reply]
Yep. Agree entirely. 

I tried to cover that in these two points:
- Button to previously mentored solutions
- Button to private notes

Anything I should flesh out in either of them?

**Comment by halfdan** [comment]
The mentoring notes are incredibly useful, but it would be really helpful if they could be shown in a sidebar instead of a modal which blocks the feedback form. Having the two side-by-side would make it a lot easier to transfer some of the mentoring advise to feedback.

**Reply by iHiD** [reply]
I like this idea. 

That would enforce the top/bottom split of code/feedback though I guess (as you wouldn't want to have `code|conversation|notes`)?

**Reply by neenjaw** [reply]
Do mentor snippets have to be private? Or at least a default public to mentors with optout? Being able to read/use them might be very helpful

**Reply by iHiD** [reply]
1. I think the mentor notes (as they are in v2) should stay public. 
2. The new "private notes" should be private to a mentor. 

But I think it would be nice if (1) was a subset of (2) effectively. Ruby has a good set up notes where that is very much the case.

**Comment by SaschaMann** [comment]
> Button to private notes

Perhaps also a button to create a PR that adds them to the public mentoring notes.

**Comment by SaschaMann** [comment]
> Ability to download the code

I don't know how this would actually work, but it would be useful to have a button that runs the command instead of just copying it. Similar to how `spotify:track:2pnZVpeKqLrEto2X9vYF6s` opens the spotify app locally. Same for the students actually.

**Reply by iHiD** [reply]
I believe this would involve needing to install an actual "app" rather than an executable. Which probably makes this a longer-term roadmap feature. But I like the idea itself.

**Reply by SaschaMann** [reply]
I took a brief look at [MSDN](https://docs.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/platform-apis/aa767914(v=vs.85)?redirectedfrom=MSDN) and it seems that you can register just about anything as URI handler. If I understand the docs correctly, it may even work with the current CLI as long as there's an install script that adds a registry entry. I'll investigate.

**Reply by SleeplessByte** [reply]
For Windows it's _extremely_ easy

**Reply by SaschaMann** [reply]
See https://github.com/exercism/cli/pull/951

**Comment by Eiermann** [comment]
It would be useful for me to have a button that toggles display of all the previous conversation on the iteration page, so I can see what suggestions I made on previous iterations without having to flip back and forth between iterations.

That would also make it easier to say things like, "check the link I included three comments ago".

**Reply by jeffdparker** [reply]
Agreed.  I always want to review the conversation to that point - on long exchanges, that takes lots of clicking.  

What about a link that takes them back to a comment, now highlighted, from a previous submission.

**Reply by iHiD** [reply]
Interesting! Maybe this is an area we need to think about more then.

**Reply by kotp** [reply]
Something like:  You have mentored this student on 3 other exercises.
Those exercise links (not published solution (in the terminology of v2)
but in the mentored mode, with conversations available to find out what
things you may have already talked about.  (Some people get really
offended if you appear venerable and repeat yourself, and sometimes
aggressive.)

**Comment by grimord** [comment]
Would it be feasible to have test execution results included alongside the solution? 

Maybe have the tests run asynchronously after the student submits their solution so as to avoid heavy load on the backend and/or delay in the submission process?

**Reply by iHiD** [reply]
Yep - these will be there, I should have mentioned. I'll add it to the main doc. Thanks :)

**Comment by workingjubilee** [comment]
I would like the ability to rejigger the layout of things somewhat. Specifically, I would like the ability to tune out automated remarks, or especially in this UI layout, replace the auto-remarks with a side-by-side of another iteration, the community mentor remarks, or my own solution.

**Reply by iHiD** [reply]
So something along the lines of a fixed LHS, with the RHS having tabs (or whatever) to allow you to see different things?

**Reply by BethanyG** [reply]
I really like this idea.  Sometimes, it is really important that I see what the student has tried or what advice I've given them.  Sometimes, I need to see the community mentor notes, or refer to my own notes from other times I've mentored.  It would be great if I could "turn on" or "turn off" areas in the RHS....or "drag" them into a priority order.  Analogous to the way some interfaces let you "anchor" or "float" different widgets.

**Reply by workingjubilee** [reply]
Yes, (relatively) fixed LHS  with a manipulable RHS through some interface. This would work less well on mobile but you could figure out alternates like tabbing between the LHS/RHS themselves, or just not since the site is fairly desktop-oriented.

**Comment by angelikatyborska** [comment]
1. 
> A mentor needs to be able to step between iterations. It is important that mentors can clearly see which iterations they have commented on previously.

It would be very helpful if there was a way to show a diff between iteration N and N-1, maybe as an alternative way to read the solution? After I ask the student to improve something and they submit a new iteration, I am more interested in the exact changes that they did than the whole solution. So far, to achieve this, I would use my IDE's "local history" feature that can show me a diff before and after downloading the new iteration.

2.
> - Button to own solution
> - Button to previously mentored solutions

I would also need a button to community solutions. I often look them up to see what are alternative ways of solving the same problem, especially when the student asks about that.

**Reply by thejoycekung** [reply]
Really love the idea of showing a diff between solutions! Particularly as students advance in iterations, their changes become smaller and it's harder to notice what's new.

**Reply by taiyab** [reply]
100% something we wanna do in general. Both for your own iterations during the process of submitting code and for your students' iterations.

**Reply by BethanyG** [reply]
Heavy plus one for diffs.  Also a plus one for links to community solutions.

**Comment by adambuckland** [comment]
Have keyboard shortcuts to these been considered? Would be nice to be able to open reference materials (such as the community notes) quickly and easily

**Reply by taiyab** [reply]
Definitely something we wanna bake in.

If people can jot down the things they'd want quick access to (next/previous iteration, etc), that would be great.

**Comment by BethanyG** [comment]
> The modal should have a CTA to improve the mentoring notes. Atm, this is a footer, but many people have said it's not clear enough.

Why not just be very up front and put an [edit on GitHub] button?  Something roughly like:

![mentor_notes](https://user-images.githubusercontent.com/5923094/89959277-e04d0780-dbf0-11ea-9df4-def84520b65b.png)

**Reply by kotp** [reply]
I know what CYA is, but not what CTA is…  is that terms to cover theirs instead our yours?  I don't get it.  And I know it isn't what you wrote, but now it has been mentioned a few times in this thread, but it seems like I am not an insider here, so I don't get the joke.

**Reply by thejoycekung** [reply]
[CTA = Call to Action](https://en.wikipedia.org/wiki/Call_to_action_(marketing)). In this case, encouraging mentors to help edit mentoring notes!

**Reply by taiyab** [reply]
@BethanyG ++

**Comment by BethanyG** [comment]
Ability to tag an interaction for discussion by other mentors -- or "bring in" another mentor for perspective?  Maybe something as simple as a pre-filled link for pasting into Slack ..or as complicated as sending a message to another mentor and flagging it in their queue?  Or maybe a "flag this for feedback on mentoring"?  Not super-clear on the exact steps -- I just know that I've benefitted greatly from getting feedback on my _mentoring_ and perspective on what a student might be saying/implying.

**Reply by iHiD** [reply]
Nice. Yes. I've previously thought about a Slack integration where you click on a solution and it pings the slack channel. But as we're getting rid of Slack (eventually), I'll probably never actually do this.

But the general idea, I like :)

**Comment by kotp** [comment]
On 20/08/11 08:05PM, Joyce Kung wrote:
> [CTA = Call to
> Action](https://en.wikipedia.org/wiki/Call_to_action_(marketing)). In
> this case, encouraging mentors to help edit mentoring notes!

Thanks, that is helpful.  Too many times acronyms are assumed to be
known, and I wasn't able to pick up on what it could mean.  I guess if I
were a marketer, I would have known that.

**Comment by SleeplessByte** [comment]
~Their handle~
~Their avatar~
~Their reputation~
~Their responsiveness.~
~If they're a favourite~
Links to previous interactions.

I don't need to see _any_ of that _whilst_ I'm mentoring. I might use it to _decide_ who I'm mentoring, and I'd expect to see it in the comments (some of it), but it's not a necessity.

Most of the time, I actually want to have their "name" which is either their first name, or their handle if their profile isn't public/they don't have one. 

-----

> A mentor needs to be able to step between iterations. It is important that mentors can clearly see which iterations they have commented on previously.

It's more important to see the _previous_ iteration _any_ mentor has commented on, and secondarily _all_ the iteration, with comments or not. But most of the time, the previous one is the only thing that really matters.

It's also important to see on which ones the _student_ has commented. Enter all the cases where a student submits 10 iterations with sometimes a comment. Right now that's really annoying to "figure out", especially if you're trying to respond to _all_ their questions.

Missing: which iterations have **unread** comments.

------

> Button to view mentor notes

I dislike this button. I hate that it's not always there. I would prefer this to be available **on the page**, or at least have a button _always_.

Right now, to get inspiration, especially if there are no mentor notes, it's:
- go to github and pray example.js is actually exemplar.js
- go to the track on non-mentor mode > the exercise > see solutions > randomly click solutions and hope for the best
- or, do the 3 dropdowns and go through your own comments of past mentors

Missing: see other solutions by other students:
- approved by me
- approved by others
- NOT approved (mentored by me) + easy way to see comments I've given there.

----

> Button to private notes

> Button to own solution

> Button to previously mentored solutions

I don't think they necessarily should be buttons.

----

> Mentors need the ability to download the students code to fiddle with it locally. This currently looks like this, but should probably be a button that shows this:

So I have to click twice? Rather not. Unless it opens the CLI automatically, then yes please!

----

> Button to mark conversation as finished

Don't get what this is supposed to be. Feels like a negative, always.

----

# Button to unapprove new iterations

and "button to "undo" approve within time limit on this iteration after approving.

**Reply by iHiD** [reply]
@SleeplessByte Two small clarifications:

> It's more important to see the previous iteration any mentor has commented on

Mentor sessions will become properly 1-1, so you'll only ever have your comments and their comments.

> Button to unapprove new iterations
> Button to mark conversation as finished

In v3, there is no approval. Instead the conversation finishes at the point that either the student or the mentor marks it as such.

**Reply by SleeplessByte** [reply]
> Mentor sessions will become properly 1-1, so you'll only ever have your comments and their comments.

What happens when I leave?

> In v3, there is no approval. Instead the conversation finishes at the point that either the student or the mentor marks it as such.

Cool. Then that's what "finished" is. Can I unfinish it?

**Comment by jeffdparker** [comment]
>  the conversation finishes at the point that either the student or the mentor marks it as such.

I would think that some students would be happy to end a conversation early, and mentors might find themselves try to get the same point across in problem after problem.

Let's take the case of the student who confuses /print/ with /return/  Do we really want to go through this with every problem?  

Or is this a straw man - perhaps the students must pass the Unit Test to advance.

**Reply by iHiD** [reply]
Thanks Jeff.

In v3, mentoring does not happen on Concept Exercises and is actively requested on Practice exercises. So mentors are only going to be interacting with students who actually want to have conversations and learn. Which I think should change the equation.

On the other hand, tests must pass for someone's Concept Exercise to be (automatically) "approved", or for a student to contact a mentor, so the print/return stuff should disappear in v3 too.

 So I hope/think that this shouldn't be a problem.

Although, I may be missing potential problems, so please feel free to challenge my logic :)
